### PR TITLE
[FLINK-6048] Asynchronous snapshots for heap-based OperatorStateBackend

### PR DESCRIPTION
**Pull request by StefanRRichter**
This PR introduces asynchronous snapshots for the heap-based `DefaultOperatorStateBackend`. Compared to the asynchronous snapshots for the heap-based keyed state backend, this implementation is rather simple and eagerly generates a deep in-memory copy of the state before running the asynchronous part of the snapshot that writes to the filesystem.

Note that this PR should later sit on top of PR #3483 and piggyback on the async-flag that was introduced. It also benefits from the improved ArrayListTypeSerializer.

Furthermore, we could have a followup that actually parallelizes checkpointing the different async backends in `AsyncCheckpointRunnable`. Previously, this was not needed because there have only been keyed state backends o those have been the only async backends.

### CODE REVIEW

**Comment by StefanRRichter** [comment]
CC @aljoscha @StephanEwen

**Comment by StefanRRichter** [comment]
Rebased.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/DefaultOperatorStateBackend.java [hunk 1]
```diff
@@ -131,59 +169,134 @@ public void dispose() {
...
+							throw new IOException("Checkpoint stream already closed.");
+						}
+					}
+
+					@Override
+					public void done(boolean canceled) {
+						if (open.compareAndSet(true, false)) {
+							CheckpointStreamFactory.CheckpointStateOutputStream stream = getIoHandle();
+							if (null != stream) {
+								closeStreamOnCancelRegistry.unregisterClosable(stream);
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/DefaultOperatorStateBackend.java]: Shouldn't we always try to remove the stream from the `ClosableRegistry` after the callable has been executed?

Comment (StefanRRichter) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/DefaultOperatorStateBackend.java]: I don't understand exactly what you mean. In line 283, the stream is unregistered. The stream is only created if the runnable is executed in which case it is guaranteed to reach `done()`. So I think whenever a stream was created and registered, it is also unregistered and closed. Or do I miss something here?

**Comment by StefanRRichter** [comment]
Thanks for the review @tillrohrmann. I addressed your comments concerning unregistration of streams. Will merge this now.

**Pull request closed without merging**